frontend/packages/dev-console/src/components/project-access/hooks.ts

```diff
@@ -0,0 +1,33 @@
+import { useK8sWatchResource } from '@console/internal/components/utils/k8s-watch-hook';
+import { ClusterRoleModel } from '@console/internal/models';
+import { K8sResourceCommon } from '@console/internal/module/k8s';
+import { defaultAccessRoles, getAvailableAccessRoles, Roles } from './project-access-form-utils';
+
+export type ClusterRoleKind = K8sResourceCommon & {
+  rules: { [key: string]: string[] }[];
+};
+
+export const useProjectAccessRoles = (): Roles => {
+  const availableClusterRoles = getAvailableAccessRoles();
+  const [clusterRoles, clusterRolesLoaded] = useK8sWatchResource<ClusterRoleKind[]>({
+    kind: ClusterRoleModel.kind,
+    isList: true,
+  });
+  if (!availableClusterRoles) {
+    return defaultAccessRoles;
+  }
+  const projectClusterRoles = clusterRoles.filter((clusterRole) => {
+    return availableClusterRoles.includes(clusterRole.metadata.name);
+  });
+  if (clusterRolesLoaded && projectClusterRoles.length === 0) {
+    return defaultAccessRoles;
+  }
+  const projectAccessRoles = projectClusterRoles.reduce((acc, projectClusterRole) => {
+    const clusterRoleName = projectClusterRole.metadata.name;
+    acc[clusterRoleName] =
+      projectClusterRole.metadata?.annotations?.['console.openshift.io/display-name'] ||
+      clusterRoleName;
+    return acc;
+  }, {});
+  return projectAccessRoles;
+};
```

Comment:
As we discussed this on slack and because I had some strange issues on my cluster, and not on a shared cluster I have done many experiments and debugging sessions in your hook. At the end I found that we have an issue with our Dropdown which doesn't support when the data (items) changed after the first rendering. (We should analyse this issue as well...)

I could not fix this with the usage of `useK8sWatchResource`. But with switch to `useK8sWatchResources` the UI works fine for me. The pro and con is that this method requested not all ClusterRoles but all configured and customized ClusterRoles individually. If the cluster admin configures no "availableClusterRoles" no API request is required.

I returned an extra `loaded` parameter so that the UI can show the LoadingBox bases on this information, so that it is not required that the hook returns an empty object for this use case (and the current Dropdown issue).

Please take a look and change whatever you want. At least on my local cluster this works fine. The fallback to the clusterRoleName works also if the API fails for any reason. (For example there is not such a ClusterRole name.)

```tsx
import * as React from 'react';
import { useK8sWatchResources } from '@console/internal/components/utils/k8s-watch-hook';
import { ClusterRoleModel } from '@console/internal/models';
import { K8sResourceCommon, WatchK8sResource } from '@console/internal/module/k8s';
import { defaultAccessRoles, getAvailableAccessRoles, Roles } from './project-access-form-utils';

export type ClusterRoleKind = K8sResourceCommon & {
  rules: { [key: string]: string[] }[];
};

export const useProjectAccessRoles = (): { data: Roles; loaded: boolean } => {
  const availableClusterRoles = getAvailableAccessRoles();

  const watchedClusterRoles = React.useMemo<Record<string, WatchK8sResource>>(() => {
    if (!availableClusterRoles || !availableClusterRoles.length) {
      return {};
    }
    return availableClusterRoles?.reduce((acc, role) => {
      acc[role] = {
        kind: ClusterRoleModel.kind,
        name: role,
      };
      return acc;
    }, {} as Record<string, WatchK8sResource>);
  }, [availableClusterRoles]);

  const clusterRoles = useK8sWatchResources<Record<string, ClusterRoleKind>>(watchedClusterRoles);

  if (!availableClusterRoles || !availableClusterRoles.length) {
    return { data: defaultAccessRoles, loaded: true };
  }

  const mappedRoles = availableClusterRoles.reduce((acc, role) => {
    const clusterRole = clusterRoles[role];
    if (clusterRole && clusterRole.loaded && clusterRole.data) {
      const label =
        clusterRole.data.metadata?.annotations?.['console.openshift.io/display-name'] ||
        clusterRole.data.metadata.name ||
        role;
      acc[role] = label;
    } else {
      acc[role] = role;
    }
    return acc;
  }, {} as Roles);

  const allLoadedOrFailed = Object.values(clusterRoles).every(
    (clusterRole) => clusterRole.loaded || clusterRole.loadError,
  );

  return { data: mappedRoles, loaded: allLoadedOrFailed };
};
```